### PR TITLE
Additional supporting functions for OCSP

### PR DESCRIPTION
**Pull request by samuel40791765**
### Issues:
Addresses `CryptoAlg-1564`

### Description of changes: 
More basic additional OCSP supporting functions.

### Call-outs:
N/A

### Testing:
New unit tests for all but `OCSP_resp_count` and `OCSP_SINGLERESP_get0_id`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and 
the ISC license.

### CODE REVIEW

File: include/openssl/ocsp.h [hunk 1]
```diff
@@ -64,6 +64,17 @@ DECLARE_ASN1_FUNCTIONS(OCSP_RESPONSE)
 DECLARE_ASN1_FUNCTIONS(OCSP_CERTID)
 DECLARE_ASN1_FUNCTIONS(OCSP_REQUEST)
 
+// d2i_OCSP_REQUEST_bio parses an OCSP request in DER format from a BIO object
+// and converts it into an |OCSP_REQUEST|.
+OPENSSL_EXPORT OCSP_REQUEST *d2i_OCSP_REQUEST_bio(BIO *bp, OCSP_REQUEST **preq);
+
+// i2d_OCSP_REQUEST_bio parses an |OCSP_REQUEST| structure and writes the OCSP
+// request in DER format into a BIO object.
+OPENSSL_EXPORT int i2d_OCSP_REQUEST_bio(BIO *bp, OCSP_REQUEST *preq);
```
Comment (andrewhop) [inline include/openssl/ocsp.h:73]:
```suggestion
// i2d_OCSP_REQUEST_bio marshals |preq| as a DER-encoded OCSPRequest and writes the result to |bp|.
OPENSSL_EXPORT int i2d_OCSP_REQUEST_bio(BIO *bp, OCSP_REQUEST *preq);
```

File: include/openssl/ocsp.h [hunk 2]
```diff
@@ -64,6 +64,17 @@ DECLARE_ASN1_FUNCTIONS(OCSP_RESPONSE)
 DECLARE_ASN1_FUNCTIONS(OCSP_CERTID)
 DECLARE_ASN1_FUNCTIONS(OCSP_REQUEST)
 
+// d2i_OCSP_REQUEST_bio parses an OCSP request in DER format from a BIO object
+// and converts it into an |OCSP_REQUEST|.
+OPENSSL_EXPORT OCSP_REQUEST *d2i_OCSP_REQUEST_bio(BIO *bp, OCSP_REQUEST **preq);
```
Comment (andrewhop) [inline include/openssl/ocsp.h:69]:
```suggestion
// d2i_OCSP_REQUEST_bio parses an OCSP request as a DER-encoded  OCSPRequest from |bp|,
// converts it into an |OCSP_REQUEST|, and writes the result  in |preq|.
OPENSSL_EXPORT OCSP_REQUEST *d2i_OCSP_REQUEST_bio(BIO *bp, OCSP_REQUEST **preq);
```

File: include/openssl/ocsp.h [hunk 3]
```diff
@@ -64,6 +64,17 @@ DECLARE_ASN1_FUNCTIONS(OCSP_RESPONSE)
...
+
+// i2d_OCSP_REQUEST_bio parses an |OCSP_REQUEST| structure and writes the OCSP
+// request in DER format into a BIO object.
+OPENSSL_EXPORT int i2d_OCSP_REQUEST_bio(BIO *bp, OCSP_REQUEST *preq);
+
+// OCSP_CERTID_dup creates a copy of an |OCSP_CERTID| structure.
+OPENSSL_EXPORT OCSP_CERTID *OCSP_CERTID_dup(OCSP_CERTID *id);
```
Comment (andrewhop) [inline include/openssl/ocsp.h:76]:
```suggestion
// OCSP_CERTID_dup allocates a new |OCSP_CERTID| and sets it equal to the state of |id|.
//  It returns the new |OCSP_CERTID| or null on error. 
OPENSSL_EXPORT OCSP_CERTID *OCSP_CERTID_dup(OCSP_CERTID *id);
```

File: crypto/ocsp/ocsp_test.cc [hunk 4]
```diff
@@ -1115,3 +1115,67 @@ TEST_P(OCSPURLTest, OCSPParseURL) {
...
+  // Create a sample OCSP_CERTID structure
+  bssl::UniquePtr<OCSP_CERTID> orig_id(
+      OCSP_cert_to_id(EVP_sha256(), subject.get(), issuer.get()));
+  ASSERT_TRUE(orig_id);
+
+  // Duplicate the OCSP_CERTID structure
+  bssl::UniquePtr<OCSP_CERTID> dup_id(OCSP_CERTID_dup(orig_id.get()));
+  ASSERT_TRUE(dup_id);
+
+  // Check that the duplicated structure has the same values as the original.
```
Comment (andrewhop) [inline crypto/ocsp/ocsp_test.cc:1168]: I realize this is just calling another ASN1 function so it's unlikely this is an issue. But for the sake of completion can you add a check that dup_id does not point to the same ocsp structure as orig_id. In theory dup could just return the same pointer and this test would still pass.

Comment (samuel40791765) [reply crypto/ocsp/ocsp_test.cc:1168]: Good point, I added an `EXPECT_NE` at the bottom.

**Review by andrewhop — APPROVED** [review]

**Review by andrewhop — APPROVED** [review]

**Review by skmcgrail — APPROVED** [review]

**Pull request merged**